tt_eager/tt_dnn/op_library/update_cache/kernels/dataflow/writer_update_cache_interleaved_start_id.cpp

```diff
@@ -38,36 +40,45 @@ void kernel_main() {
-            cb_wait_front(untilized_cache_cb_id, Wt);
-            cb_reserve_back(untilized_cache2_cb_id, Wt);
-            uint32_t cache_l1_write_addr = get_read_ptr(untilized_cache_cb_id) + offset;
-            noc_async_read(input_l1_read_addr, cache_l1_write_addr, Wbytes);
-            noc_async_read_barrier();
-            cb_push_back(untilized_cache2_cb_id, Wt);
-            input_l1_read_addr += Wbytes;
-            // Compute will pop both cache cbs together
-
-            cb_wait_front(cache_cb_id, Wt);
-            uint32_t out_l1_read_addr = get_read_ptr(cache_cb_id);
-            for(uint32_t curr_cache_id = cache_id; curr_cache_id < cache_id + Wt; ++curr_cache_id) {
-                noc_async_write_tile(curr_cache_id, s0, out_l1_read_addr);
-                out_l1_read_addr += cache_tile_bytes;
+        for (uint32_t u = 0; u < u_count; ++u) {
+            // Operating on a granularity > 1 led to performance improvements.
+            // It introduces a double-buffered pipeline between compute and writer.
+            for (uint32_t g = 0; g < granularity; ++g) {
+                // Wait on compute to untilize a block. Update that block in L1.
+                cb_wait_front(untilized_cache_cb_id, Wt);
+                cb_reserve_back(untilized_cache2_cb_id, Wt);
+                uint32_t cache_l1_write_addr = get_read_ptr(untilized_cache_cb_id) + offset;
+                noc_async_read(input_l1_read_addr, cache_l1_write_addr, Wbytes);
+                input_l1_read_addr += Wbytes;
+                noc_async_read_barrier();
+                cb_push_back(untilized_cache2_cb_id, Wt);
+                cb_pop_front(untilized_cache_cb_id, Wt); // NEW
             }
-            cache_id += cache_batch_num_tiles; // Input is read in by batch, then heads so skip to next batch
-            b++;
-            if (b == B) {
-                b = 0;
-                cache_id = cache_id - cache_total_num_tiles + cache_head_num_tiles; // Start of next head
+
+
+            for (uint32_t g = 0; g < granularity; ++g) {
+                // Wait on compute to tilize an updated block. Write that block to DRAM
+                cb_wait_front(cache_cb_id, Wt);
+                uint32_t out_l1_read_addr = get_read_ptr(cache_cb_id);
+                for(uint32_t curr_cache_id = cache_id; curr_cache_id < cache_id + Wt; ++curr_cache_id) {
+                    noc_async_write_tile(curr_cache_id, s0, out_l1_read_addr);
+                    out_l1_read_addr += cache_tile_bytes;
+                }
+                cache_id += cache_batch_num_tiles; // Input is read in by batch, then heads so skip to next batch
+                b++;
+                if (b == B) {
+                    b = 0;
+                    cache_id = cache_id - cache_total_num_tiles + cache_head_num_tiles; // Start of next head
+                }
+                cb_pop_front(cache_cb_id, Wt);
             }
+            // Delay syncing the writes to maximize perf.
```

Comment:
Isn't it somewhat unsafe to pop the cbs before write barrier or write flush? Data could potentially be written in now that the cb has free space (seems highly unlikely in this setup though)

Reply:
I think the barrier location is fine but pops should still be after. For moving the cbs you can use the flush barrier here instead of write barrier, and maybe a write barrier at the very end?

Reply:
Thanks Austin! I moved the aync_write_barrier back to its original location because otherwise non deterministic behaviour was occuring.

Reply:
Yeah Austin, I'm suspicious of this too.
I made it so that the output CB can buffer a full batch of tiles. The compute should write into position 0 .. B before the CB wraps back to 0. I think that's why it works.

What I really wanted to do was reserve inside that loop, sync at the end of the loop, and then pop however many times is necessary. However, I saw bad PCC with that. The docs have strange warnings about calling wait_front multiple times without matching pop_front calls.

Reply:
I was syncing offline with Reem and I think you can do this.
```
cb_wait(granularity * Wt);
get_ptr;
loop granularity and update ptr each loop;
write_barrier();
cb_pop(granularity * Wt);
```
Since reader is pushing granularity * Wt at a time anyways, there shouldnt be a perf impact for waiting for this amount instead of just Wt

Reply:
Hey guys, I used this system instead:
```
                noc_async_writes_flushed();
                cb_pop_front(cache_cb_id, Wt);
            }
            }
            noc_async_write_barrier();
            cb_pop_front(cache_cb_id, Wt);
        }
        }
        cb_pop_front(untilized_input_cb_id, Wt);
        cb_pop_front(untilized_input_cb_id, Wt);
    }
    }
    // Delay syncing the writes to maximize perf.
    noc_async_write_barrier();
}
}
```
with flushing the writes, and I get back to the perf gain that we initially got. @cglagovich can you confirm on your side?